### ISSUE TITLE
WinCNG test failures

### ISSUE THREAD

**Issue by dfandrich** [libssh2/libssh2#804]
Most runs, the WinCNG openssh test runs on Appveyor fail at least one test. The actual test changes (e.g. test_password_auth_fails_with_wrong_username, test_keyboard_interactive_auth_fails_with_wrong_response, test_hostkey) and sometimes the WinCNG runs complete without failures, but it's pretty common that at least one fails. It's often on the 
VS2015, WinCNG, Shared, x86_64 build (but maybe because that's one of the first WinCNG builds and the tests stop after the first error), but it happens on other WinCNG builds, too, including VS2013, Static and x86 builds, but always with WinCNG.

The tests fail with the error `9: libssh2_session_handshake failed (-8): Unable to exchange encryption keys`. Running the tests with `FIXTURE_TRACE_ALL=1` showed this log at the time of failure:

```
[00:05:13] 5: [libssh2] 0.951417 Transport: Packet type 33 received, length=1078
[00:05:13] 5: [libssh2] 0.951417 Transport: Looking for packet of type: 33
[00:05:13] 5: [libssh2] 0.951417 Key Ex: Server's MD5 Fingerprint: 0c:0e:d1:a5:bb:10:27:5f:76:92:4c:e1:87:ce:5c:5e
[00:05:13] 5: [libssh2] 0.951417 Key Ex: Server's SHA1 Fingerprint: f3:cd:59:e2:91:3f:44:22:b8:0f:7b:0a:82:b2:b8:9e:ae:44:93:87
[00:05:13] 5: [libssh2] 0.951417 Key Ex: Server's SHA256 Fingerprint: kuPaSd88f5moKPUF7YI5OXpdH2KRRFl2D4ePdRD1Y6M=
[00:05:13] 5: [libssh2] 1.045165 Failure Event: -11 - Unable to verify hostkey signature
[00:05:13] 5: [libssh2] 1.045165 Failure Event: -8 - Unrecoverable error exchanging keys
[00:05:13] 5: [libssh2] 1.045165 Failure Event: -8 - Unable to exchange encryption keys
[00:05:13] 5: libssh2_session_handshake failed (-8): Unable to exchange encryption keys
[00:05:13] 5: [libssh2] 1.045165 Transport: Disconnecting: reason=11, desc=test ended, lang=
[00:05:13] 5: => libssh2_transport_write plain (23 bytes)
[00:05:13] 5: 0000: 01 00 00 00 0B 00 00 00  0A 74 65 73 74 20 65 6E : .........test en
[00:05:13] 5: 0010: 64 65 64 00 00 00 00                             : ded....
```

I  wouldn't expect to see this kind of problem (hostkey verification problem) if it were a timing issue. Could it be a reliance on uninitialized memory? It would be interesting to run this on the Windows equivalent to Valgrind or ASAN/UBSAN to see if something shows up.

**Comment by vszakats** [comment]
Here's a live run reproducing this failure, with trace enabled: https://ci.appveyor.com/project/libssh2org/libssh2/builds/46822855/job/c3fw767k5bm3sgjf?fullLog=true#L19325

**Comment by mback2k** [comment]
I have observed this before and think the root cause is the fact that we use [random bytes to generate the DH session key](https://github.com/libssh2/libssh2/blob/17801d206444a234c25084edb01e26485e130bcf/src/wincng.c#L2143) and then stuff those bytes which are not necessarily an actual prime into the [RSA function to perform the modular exponentiation](https://github.com/libssh2/libssh2/blob/17801d206444a234c25084edb01e26485e130bcf/src/wincng.c#L2218). I based my implementation of the random bignum function for WinCNG on other backends that like mbedTLS which do it similarly.

**Comment by vszakats** [comment]
@mback2k Looks like a good lead. After eyeballing these functions closely, I could not find an obvious coding error and the issue applies not just to rsa-sha2, but also to rsa-sha1 (confirming this isn't a fresh regression.)

Do you have idea for a fix?

**Comment by vszakats** [comment]
Agreed on the value of Vista-compatibility. This issue is happening also with the Visual Studio 2022 AppVeyor image, with Windows Server 2019 (Windows 10), what could be an explanation?

Three such runs with some logging details (I've since add logging for `BCRYPT_DH_ALGORITHM` detection):
https://ci.appveyor.com/project/libssh2org/libssh2/builds/46828624/job/0pto6u8d1hbxokb9
https://ci.appveyor.com/project/libssh2org/libssh2/builds/46827388/job/4e1ivdrjeepp1g8c
https://ci.appveyor.com/project/libssh2org/libssh2/builds/46826769/job/v4rrlp9vvsfhcioy

**Comment by vszakats** [comment]
Tried to clear allocated buffers (as part of the warning-silencing PR), but it didn't help. Tried tweaking internals to use the modern Crypto features of Windows, that also didn't result in success. Tried disabling RSA, which broke tests as WinCNG doesn't support ECDA and ED25519 hostkey, so added DSA hostkeys, but couldn't make this work.

I'm proposing to disable WinCNG tests till we find a solution. It causes almost all CI runs to go red.

What do you think?

**Comment by vszakats** [comment]
This has been mitigated in our CI tests by retrying in case of a KEX failure: bc120a343bb3cda0985fb06ee491ec5318e98426

This helps getting realistic CI results (greens instead of always-red) and also helps to avoid this issue masking others.

**Comment by vszakats** [comment]
Potentially related server-side workaround by setting `KexAlgorithms -diffie-hellman-group-exchange-sha25`:
https://github.com/logmanoriginal/lvssh2/commit/ae7c1aa21cf564695ff0045cf9d9a60ad2680e24

"tests: Fix unable to exchange encryption keys
There appears to be a bug in libssh2 when using the WINCNG crypto
backend, which results in error -8 "Unable to exchange encryption
keys" during key exchange. The error happens during diffie-hellman
key-pair generation ("dh key pair generation failed") with method
'diffie-hellman-group-exchange-sha256'.

This appears to be a known issue that may be addressed in future
versions of libssh2: https://github.com/libssh2/libssh2/issues/1370

The workaround is to disable the offending method in sshd_config."

```diff
diff --git a/tests/docker/openssh-server/config/ssh_host_keys/sshd_config b/tests/docker/openssh-server/config/ssh_host_keys/sshd_config
index 2eba339..85ce834 100644
--- a/tests/docker/openssh-server/config/ssh_host_keys/sshd_config
+++ b/tests/docker/openssh-server/config/ssh_host_keys/sshd_config
@@ -119,3 +119,8 @@ Subsystem	sftp	internal-sftp
 #	AllowTcpForwarding no
 #	PermitTTY no
 #	ForceCommand cvs server
+
+# Disable diffie-hellman-group-exchange-sha256 to prevent
+# error "dh key pair generation failed" when using libssh2
+# with WINCNG. This is a workaround for libssh2 bug #1370
+KexAlgorithms -diffie-hellman-group-exchange-sha256
```

```
Key Ex: Agreed on KEX method: diffie-hellman-group-exchange-sha256
...
Failure Event: -5 - dh key pair generation failed
Failure Event: -8 - Unrecoverable error exchanging keys
Failure Event: -8 - Unable to exchange encryption keys
```

via https://github.com/libssh2/libssh2/issues/1370#issuecomment-2143477231

**Comment by mback2k** [comment]
Good find, but finding the root cause would still be interesting.
